include/tscore/ArgParser.h

```diff
@@ -0,0 +1,255 @@
+/** @file
+
+  Powerful and easy-to-use command line parsing for ATS
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <map>
+#include <vector>
+#include <functional>
+#include <string_view>
+
+// more than zero arguments
+constexpr unsigned MORE_THAN_ZERO_ARG_N = ~0;
+// more than one arguments
+constexpr unsigned MORE_THAN_ONE_ARG_N = ~0 - 1;
+// customizable indent for help message
+constexpr int INDENT_ONE   = 24;
+constexpr int INDENT_TWO   = 32;
+constexpr int INDENT_THREE = 46;
+
+namespace ts
+{
+using AP_StrVec = std::vector<std::string>;
+// The class holding both the ENV and String arguments
+class ArgumentData
+{
+public:
+  // bool to check if certain command/option is called
+  operator bool() const noexcept { return _is_called; }
+  // index accessing []
+  std::string const &operator[](int x) const { return _values.at(x); }
+  // return the Environment variable
+  std::string const &env() const noexcept;
+  // iterator for arguments
+  AP_StrVec::const_iterator begin() const noexcept;
+  AP_StrVec::const_iterator end() const noexcept;
+  // index accessing
+  std::string const &at(unsigned index) const;
+  // access the first index, equivalent to at(0)
+  std::string const &value() const noexcept;
```

Comment:
What happens if you call this when empty() == true?

Reply:
it will return an empty string.